Change jira-issue-processing workflow

fehnkera opened the pull request:
The issue extraction in 'jira_issue_processing.py' now processes the xml-files successively instead as previously load and process all xml-files together. Therefore output-files get cleared at the beginning and results get appended during  the processing.
    
Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

Commit:
Addition to Pull-Request #16

Element is printed to the result file as string but should be a list.
Corrects this.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

Commit:
Change workflow of 'jira_issue_processing.py'

Now the issue extraction processes the xml-files successively instead as
previously load and process all xml-files together. Therefore
output-files get cleared at the begining and results get appended during
the processing.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

Commit:
Update copyright header

The copyright header in 'csv_writer.py' is updated.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

clhunsen commented on csv_writer/csv_writer.py:
```diff
@@ -34,13 +35,28 @@ def __encode(line):
+    :param append: Flag if lines shall be appended to file or overwrite file
+    """
+
+    if append:
```
Why don't you just use the switch for the second argument of `open`? This would remove the code duplication here. Something like:
```python
open_mode = "a+b" if append else "wb"
```

To give some more explanation on this idea: I used `a+b` in the example above as `a+` [ensures that the file is created if not existent before](https://docs.python.org/2.7/library/functions.html#open), while `b` ensures compatibility with the `wb` in the other case of the if-statement.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -70,52 +70,82 @@ def run():
-    print_to_disk_bugs(issues, __resdir)
+
+    # load the xml-file list
+    file_list = [f for f in os.listdir(__srcdir) if os.path.isfile(os.path.join(__srcdir, f))]
```
@ecklbarb: Out of curiosity: Can there be the case that not only XML files are in this folder?

ecklbarb replied:
There are usually only XML files in the folder. But we had already the situation where XML files were broken and couldn't be read. @clhunsen, can we check the correctness of a XML file?

clhunsen replied:
Honestly, I do not have clue right now. I guess, when you call the parser on a malformed or non-existent XML file, it will throw an exception. At least, this is worth a try.

If my assumption is true, the handling is easy: We handle the exceptions in the function `load_xml`, return `None` in the error case, and handle this case then in the outer for-loop by skipping the current file (while giving a log message!).

ecklbarb replied:
Your guess is right. If the file can't be read, we get an exception and it breaks off.

I think, skipping a file is not a good idea, because we get incomplete data. I suggest that the user can decide what to do. At the beginning, I ignore the missing data and delete the malformed file. Now, I generate the xml-files again, and hope to get complete data or I merge the two data sets to one without malformed files.

clhunsen replied:
By skipping a malformed file, I did not mean to ignore it. Actually, with handling an exception, you gain two advantages: First, you get incremental (but well-formed) data to work with in further analysis steps. Second, you can print a complete list of malformed files at the end – without running the extraction process several times, each time waiting for the next error.

clhunsen commented on csv_writer/csv_writer.py:
```diff
@@ -34,13 +35,28 @@ def __encode(line):
-            wr.writerow(line_encoded)
+def write_to_csv(file_path, lines, append=False):
+    """
+    Write the given lines to the file with the given file path.
```
Please insert an empty line here. But, above all, thank you for the documentation! :+1:

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -70,52 +70,82 @@ def run():
+    clear_result_files(__resdir)
+
+    # processes every xml-file
+    for file in file_list:
```
Just noticing now: `file` is a reserved word in Python! To avoid any problems with that, can you please change the loop variable for something else? E.g., `current_file`.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -70,52 +70,82 @@ def run():
+    csv_writer.write_to_csv(output_file_bugs, [])
+    csv_writer.write_to_csv(output_file_extr, [])
+    csv_writer.write_to_csv(output_file_gephi_edges, [])
+    csv_writer.write_to_csv(output_file_gephi_nodes, [])
```
Is there a reason why you do not use a list for the files and for-loop for the file creation/truncation here?

For example, `open(result_file, "w+").close()` would either create a non-existent file `result_file` or truncate it if it exists already (i.e., remove all content). This would make a loop-version very concise and short.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -70,52 +70,82 @@ def run():
+    csv_writer.write_to_csv(output_file_gephi_nodes, [])
+
+
+def load_xml(source_folder, file):
```
Same as above: Please change the name of the second parameter into `xml_file`, for example.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -645,7 +676,7 @@ def print_to_disk_bugs(issues, results_folder):
 
     # write to output file
-    csv_writer.write_to_csv(output_file, lines)
+    csv_writer.write_to_csv(output_file, lines, True)
```
Please add the name for the last parameter as it is not self-explaining: `append=True`.

This also holds for all other occurrences below.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -70,52 +70,82 @@ def run():
-def load_xml(source_folder):
+def clear_result_files(results_folder):
+    """
+    Creates an empty csv file for every result file.
```
Please insert an empty line here.

clhunsen reviewed (changes requested):
Thank you for these changes, @fehnkera! This will really improve the workflow and also the debuggability. :+1: 

I commented on some minor things to fix. Though, there are two places where we should discuss how to reduce code duplication (I already added some ideas to my comments).

Commit:
Apply review and simplify code

The review from @clhunsen and @ecklbarb is applied. This incluedes
codestyle changes, removance of duplicated code aswell as
exception-handling for malformed data. Also the return type of
'load_xml' changes from a single-element-list to just the element. That
way a for-loop in 'parse_xml' is not needed anymore.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

clhunsen reviewed (approved):
Thanks for the additional enhancements, @fehnkera! Unless @ecklbarb has any objections, we can merge later.
